atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java

```diff
@@ -652,6 +633,97 @@ public String toString() {
         }
     }
 
+    private Map<Cell, Value> fetchAllCellsForRows(final InetSocketAddress host,
+            final TableReference tableRef,
+            final List<byte[]> rows,
+            final long startTs) throws Exception {
+
+        final Map<ByteBuffer, List<ColumnOrSuperColumn>> result = Maps.newHashMap();
+        Map<ByteBuffer, List<ColumnOrSuperColumn>> partialResult;
+        List<KeyPredicate> query = cellLoader.translateRowsToKeyPredicates(rows, config.fetchPerRowReadLimit());
+
+        //todo(Sudiksha): refactor
+        while (!query.isEmpty()) {
+
+            partialResult = fetchCellsForKeyPredicates(host, tableRef, query, startTs);
+            query.clear();
+
+            //todo refactor
+            for(Entry<ByteBuffer, List<ColumnOrSuperColumn>> cellsForRow: partialResult.entrySet()) {
+
+                List<ColumnOrSuperColumn> cells = cellsForRow.getValue();
+
+                if (!cells.isEmpty()) {
+                    ByteBuffer row = cellsForRow.getKey();
+                    result.getOrDefault(row, new ArrayList<>()).addAll(cells);
+                    query.add(new KeyPredicate()
```

Comment:
this feels oddly similar to `translateRowsToKeyPredicates`

Reply:
perhaps instead of mutating the `query`, how about just setting it at the bottom? it would be nice to just make a new query object each time but we probably don't want recursion here.. so need some mutation somewhere.

Reply:
that way, I *think* you can reuse `translateRowsToKeyPredicates`, although I may be mistaken